File: helpers/parser/parserUtils.js
const digraphs = ["k", "c", "ṭ", "t", "p", "g", "j", "ḍ", "d", "b"]
  .concat(["k", "c", "ṭ", "t", "p", "g", "j", "ḍ", "d", "b"].map(d => d.toUpperCase()));

const vowels = ["a", "i", "u", "e", "ā", "ī", "ū", "o"]
  .concat(["a", "i", "u", "e", "ā", "ī", "ū", "o"].map(v => v.toUpperCase()));

const long = ["e", "ā", "ī", "ū", "o"]
  .concat(["e", "ā", "ī", "ū", "o"].map(l => l.toUpperCase()));

const short = ["a", "i", "u"]
  .concat(["a", "i", "u"].map(s => s.toUpperCase()));

const consonants = [
  "k", "c", "ṭ", "t", "p", "kh", "ch", "ṭh", "th", "ph", "gh", "jh", "ḍh", "dh", "bh",
  "g", "j", "ḍ", "d", "b", "ṅ", "ñ", "ṇ", "n", "m", "h", "y", "r", "l", "v", "ś", "ṣ", "s", "ṁ", "ṃ"
]
  .concat([
    "k", "c", "ṭ", "t", "p", "kh", "ch", "ṭh", "th", "ph", "gh", "jh", "ḍh", "dh", "bh",
    "g", "j", "ḍ", "d", "b", "ṅ", "ñ", "ṇ", "n", "m", "h", "y", "r", "l", "v", "ś", "ṣ", "s", "ṁ", "ṃ"
  ].map(c => c.toUpperCase()))
  .concat(["Kh", "Ch", "Ṭh", "Th", "Ph", "Gh", "Jh", "Ḍh", "Dh", "Bh"]);

function tokenize(line) {
  let tokens = [];
  let i = 0;
  while (i < line.length) {
    let char = line[i];
    if (digraphs.includes(char) && i + 1 < line.length && line[i + 1] === 'h') {
      tokens.push(char + 'h');
      i += 2;
    } else {
      tokens.push(char);
      i++;
    }
  }
  return tokens;
}

function findLastVowelIndex(tokens) {
  for (let i = tokens.length - 1; i >= 0; i--) {
    if (vowels.includes(tokens[i])) {
      return i;
    }
  }
  return -1;
}

function processTokens(tokens, heavyEnd = false, useXBelow = false) {
  let processedTokens = [];
  const lastVowelIndex = heavyEnd ? findLastVowelIndex(tokens) : -1;
  
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];
    if (long.includes(token)) {
      if (heavyEnd && i === lastVowelIndex) {
        if (useXBelow) {
          processedTokens.push(token + '\u0353');
        } else {
          processedTokens.push(token + '\u0331');
        }
      } else {
        processedTokens.push(token + '\u0331');
      }
    } else if (short.includes(token)) {
      if (heavyEnd && i === lastVowelIndex) {
        if (useXBelow) {
          processedTokens.push(token + '\u0353');
        } else {
          processedTokens.push(token + '\u0331');
        }
      } else {
        const [n1, n2] = findNextTokens(tokens, i);
        if (n1 === ' ' || n1 === null || (consonants.includes(n1) && vowels.includes(n2)) || (n1 === 'b' && n2 === 'r') || (n1 === 'b' && n2 === 'y') || (n1 === 'v' && n2 === 'y')) {
          processedTokens.push(token + '\u032E');
        } else {
          processedTokens.push(token + '\u0331');
        }
      }
    } else {
      processedTokens.push(token);
    }
  }
  return processedTokens;
}

function findNextTokens(tokens, currentIndex) {
  let n1 = null;
  let n2 = null;
  let count = 0;
  for (let i = currentIndex + 1; i < tokens.length; i++) {
    const token = tokens[i];
    if (consonants.includes(token) || vowels.includes(token) || token === ' ') {
      if (count === 0) {
        n1 = token;
      } else if (count === 1) {
        n2 = token;
        break;
      }
      count++;
    }
  }
  return [n1, n2];
}

module.exports = {
  tokenize,
  processTokens,
};

File: helpers/parser_py/parser_utils.py
digraphs = ["k", "c", "ṭ", "t", "p", "g", "j", "ḍ", "d", "b"] + [c.upper() for c in ["k", "c", "ṭ", "t", "p", "g", "j", "ḍ", "d", "b"]]

vowels = ["a", "i", "u", "e", "ā", "ī", "ū", "o"] + [v.upper() for v in ["a", "i", "u", "e", "ā", "ī", "ū", "o"]]

long = ["e", "ā", "ī", "ū", "o"] + [l.upper() for l in ["e", "ā", "ī", "ū", "o"]]

short = ["a", "i", "u"] + [s.upper() for s in ["a", "i", "u"]]

consonants = [
  "k", "c", "ṭ", "t", "p", "kh", "ch", "ṭh", "th", "ph", "gh", "jh", "ḍh", "dh", "bh",
  "g", "j", "ḍ", "d", "b", "ṅ", "ñ", "ṇ", "n", "m", "h", "y", "r", "l", "v", "ś", "ṣ", "s", "ṁ", "ṃ"
] + [c.upper() for c in [
    "k", "c", "ṭ", "t", "p", "kh", "ch", "ṭh", "th", "ph", "gh", "jh", "ḍh", "dh", "bh",
    "g", "j", "ḍ", "d", "b", "ṅ", "ñ", "ṇ", "n", "m", "h", "y", "r", "l", "v", "ś", "ṣ", "s", "ṁ", "ṃ"
]] + ["Kh", "Ch", "Ṭh", "Th", "Ph", "Gh", "Jh", "Ḍh", "Dh", "Bh"]

def tokenize(line):
    tokens = []
    i = 0
    while i < len(line):
        char = line[i]
        if char in digraphs and i + 1 < len(line) and line[i + 1] == 'h':
            tokens.append(char + 'h')
            i += 2
        else:
            tokens.append(char)
            i += 1
    return tokens

def find_last_vowel_index(tokens):
    for i in range(len(tokens) - 1, -1, -1):
        if tokens[i] in vowels:
            return i
    return -1

def process_tokens(tokens, heavy_end=False, use_x_below=False):
    processed_tokens = []
    last_vowel_index = find_last_vowel_index(tokens) if heavy_end else -1

    for i, token in enumerate(tokens):
        if token in long:
            if heavy_end and i == last_vowel_index:
                processed_tokens.append(token + ('\u0353' if use_x_below else '\u0331'))
            else:
                processed_tokens.append(token + '\u0331')
        elif token in short:
            if heavy_end and i == last_vowel_index:
                processed_tokens.append(token + ('\u0353' if use_x_below else '\u0331'))
            else:
                n1, n2 = find_next_tokens(tokens, i)
                if n1 == ' ' or n1 is None or (n1 in consonants and n2 in vowels) or (n1 == 'b' and n2 == 'r') or (n1 == 'b' and n2 == 'y') or (n1 == 'v' and n2 == 'y'):
                    processed_tokens.append(token + '\u032E')
                else:
                    processed_tokens.append(token + '\u0331')
        else:
            processed_tokens.append(token)
    return processed_tokens

def find_next_tokens(tokens, current_index):
    n1 = None
    n2 = None
    count = 0
    for i in range(current_index + 1, len(tokens)):
        token = tokens[i]
        if token in consonants or token in vowels or token == ' ':
            if count == 0:
                n1 = token
            elif count == 1:
                n2 = token
                break
            count += 1
    return n1, n2
